lib/internal/modules/esm/resolve.js

```diff
@@ -466,17 +469,34 @@ function resolvePackageTargetString(
 
   const composeResult = (resolved) => {
     let format;
-    try {
-      format = getPackageType(resolved);
-    } catch (err) {
-      if (err.code === 'ERR_INVALID_FILE_URL_PATH') {
-        const invalidModuleErr = new ERR_INVALID_MODULE_SPECIFIER(
-          resolved, 'must not include encoded "/" or "\\" characters', base);
-        invalidModuleErr.cause = err;
-        throw invalidModuleErr;
+
+    const ext = extname(resolved.pathname);
+    if (ext === '.js') {
+      try {
+        format = getPackageType(resolved);
+      } catch (err) {
+        if (err.code === 'ERR_INVALID_FILE_URL_PATH') {
+          const invalidModuleErr = new ERR_INVALID_MODULE_SPECIFIER(
+            resolved, 'must not include encoded "/" or "\\" characters', base);
+          invalidModuleErr.cause = err;
+          throw invalidModuleErr;
+        }
+        throw err;
       }
-      throw err;
+    } else {
+      format = extensionFormatMap[ext];
     }
+
+    if (format == null && experimentalSpecifierResolution === 'node') {
+      if (!experimentalSpecifierResolutionWarned) {
+        process.emitWarning(
+          'The Node.js specifier resolution in ESM is experimental.',
+          'ExperimentalWarning');
+        experimentalSpecifierResolutionWarned = true;
+      }
+      format = legacyExtensionFormatMap[ext];
```

Comment:
Would it be clearer to define a function that resolves the legacy format and emit the warning?
```suggestion
      format = extensionFormatMap[ext] ?? getLegacyExtensionFormat(ext);
```